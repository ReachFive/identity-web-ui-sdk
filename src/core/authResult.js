import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
