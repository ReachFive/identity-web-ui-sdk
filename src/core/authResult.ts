import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
