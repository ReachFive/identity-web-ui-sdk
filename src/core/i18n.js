import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
