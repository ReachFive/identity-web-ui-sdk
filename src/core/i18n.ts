export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
