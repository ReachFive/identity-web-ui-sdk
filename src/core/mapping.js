import { getValue, setValue } from '../helpers/propertyHelpers';

export class PathMapping {
    constructor(modelPath) {
        this.modelPath = modelPath;
    }

    bind(model) {
        return getValue(model, this.modelPath);
    }

    unbind(model, value) {
        return setValue(model, this.modelPath, value);
    }
}
