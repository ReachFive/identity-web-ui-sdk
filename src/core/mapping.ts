import { ProfileAddress } from '@reachfive/identity-core';
import { getValue, setValue } from '../helpers/propertyHelpers';

export interface PathMapping {
    bind<T extends Record<string, unknown>>(model: T): unknown;
    unbind<T extends Record<string, unknown>, V>(model: T, value: V): T | V;
}

export class DefaultPathMapping implements PathMapping {
    protected readonly modelPath: string;

    constructor(modelPath: string) {
        this.modelPath = modelPath;
    }

    bind<T extends Record<string, unknown>>(model: T) {
        return getValue(model, this.modelPath);
    }

    unbind<T extends Record<string, unknown>, V>(model: T, value: V) {
        return setValue(model, this.modelPath, value);
    }
}

export class AddressPathMapping implements PathMapping {
    protected readonly property: string;

    constructor(property: string) {
        this.property = property;
    }

    private defaultIndex(model: { addresses?: ProfileAddress[] }) {
        const defaultIndex = (model.addresses ?? []).findIndex(address => address.isDefault);
        return defaultIndex > -1 ? defaultIndex : 0;
    }

    bind<T extends { addresses?: ProfileAddress[] }>(model: T) {
        return getValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`);
    }

    unbind<T extends { addresses?: ProfileAddress[] }, V>(model: T, value: V) {
        return setValue(model, `addresses.${this.defaultIndex(model)}.${this.property}`, value);
    }
}
