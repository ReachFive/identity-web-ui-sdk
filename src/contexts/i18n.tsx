import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
