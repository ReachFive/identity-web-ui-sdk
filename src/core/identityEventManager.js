import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
