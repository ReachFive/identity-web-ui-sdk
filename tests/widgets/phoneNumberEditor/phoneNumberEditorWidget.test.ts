/**
 * @jest-environment jest-fixed-jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom/jest-globals'
import 'jest-styled-components';

import { type Client } from '@reachfive/identity-core';

import { type I18nMessages } from '../../../src/core/i18n';

import PhoneNumberEditorWidget from '../../../src/widgets/phoneNumberEditor/phoneNumberEditorWidget';
import { componentGenerator, snapshotGenerator } from '../renderer';

const defaultI18n: I18nMessages = {}

describe('Snapshot', () => {
    // @ts-expect-error partial Client
    const apiClient: Client = {
        updatePhoneNumber: jest.fn<Client['updatePhoneNumber']>().mockResolvedValue()
    }

    const generateSnapshot = snapshotGenerator(PhoneNumberEditorWidget, apiClient, defaultI18n)

    describe('phone number editor', () => {
        test('basic',
            generateSnapshot({ accessToken: 'azerty' })
        );
    });
})

describe('DOM testing', () => {
    const updatePhoneNumber = jest.fn<Client['updatePhoneNumber']>()
    const verifyPhoneNumber = jest.fn<Client['verifyPhoneNumber']>()

    const onError = jest.fn()
    const onSuccess = jest.fn()
    
    beforeEach(() => {
        updatePhoneNumber.mockClear()
        onError.mockClear()
        onSuccess.mockClear()
    })

    // @ts-expect-error partial Client
    const apiClient: Client = {
        updatePhoneNumber,
        verifyPhoneNumber,
    }

    const generateComponent = componentGenerator(PhoneNumberEditorWidget, apiClient, defaultI18n)

    describe('phoneNumberEditor', () => {
        test('default', async () => {
            const user = userEvent.setup()

            updatePhoneNumber.mockResolvedValue()
            verifyPhoneNumber.mockResolvedValue()

            await generateComponent({ accessToken: 'azerty', onError, onSuccess })

            const phoneNumberInput = screen.getByLabelText('phoneNumber')
            expect(phoneNumberInput).toBeInTheDocument()
            
            await userEvent.clear(phoneNumberInput)
            await userEvent.type(phoneNumberInput, '+33123456789')
            
            const submitBtn = screen.getByRole('button', { name: 'send'})
            expect(submitBtn).toBeInTheDocument()

            await user.click(submitBtn)

            expect(updatePhoneNumber).toBeCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    phoneNumber: '+33123456789',
                })
            )

            const verificationCodeInput = screen.getByLabelText('verificationCode')
            expect(verificationCodeInput).toBeInTheDocument()
            
            await userEvent.clear(verificationCodeInput)
            await userEvent.type(verificationCodeInput, '123456')
            
            const submitCodeBtn = screen.getByRole('button', { name: 'send'})
            expect(submitCodeBtn).toBeInTheDocument()

            await user.click(submitCodeBtn)

            expect(verifyPhoneNumber).toBeCalledWith(
                expect.objectContaining({
                    accessToken: 'azerty',
                    phoneNumber: '+33123456789',
                    verificationCode: '123456',
                })
            )

            expect(onSuccess).toBeCalled()
            expect(onError).not.toBeCalled()
        })

        test('api update phoneNumber failed', async () => {
            const user = userEvent.setup()

            updatePhoneNumber.mockRejectedValue(new Error('Unexpected error'))

            await generateComponent({ accessToken: 'azerty', onError, onSuccess })

            const phoneNumberInput = screen.getByLabelText('phoneNumber')
            expect(phoneNumberInput).toBeInTheDocument()
            
            await userEvent.clear(phoneNumberInput)
            await userEvent.type(phoneNumberInput, '+33123456789')
            
            const submitBtn = screen.getByRole('button', { name: 'send'})

            await user.click(submitBtn)

            expect(onSuccess).not.toBeCalled()
            expect(onError).toBeCalledWith(new Error('Unexpected error'))
        })

        test('api code verificationfailed', async () => {
            const user = userEvent.setup()

            updatePhoneNumber.mockResolvedValue()
            verifyPhoneNumber.mockRejectedValue(new Error('Unexpected error'))

            await generateComponent({ accessToken: 'azerty', onError, onSuccess })

            const phoneNumberInput = screen.getByLabelText('phoneNumber')
            expect(phoneNumberInput).toBeInTheDocument()
            
            await userEvent.clear(phoneNumberInput)
            await userEvent.type(phoneNumberInput, '+33123456789')
            
            const submitBtn = screen.getByRole('button', { name: 'send'})
            expect(submitBtn).toBeInTheDocument()

            await user.click(submitBtn)

            expect(updatePhoneNumber).toBeCalled()

            const verificationCodeInput = screen.getByLabelText('verificationCode')
            expect(verificationCodeInput).toBeInTheDocument()
            
            await userEvent.clear(verificationCodeInput)
            await userEvent.type(verificationCodeInput, '123456')
            
            const submitCodeBtn = screen.getByRole('button', { name: 'send'})

            await user.click(submitCodeBtn)

            expect(verifyPhoneNumber).toBeCalled()

            expect(onSuccess).not.toBeCalled()
            expect(onError).toBeCalledWith(new Error('Unexpected error'))
        })
    })

})
