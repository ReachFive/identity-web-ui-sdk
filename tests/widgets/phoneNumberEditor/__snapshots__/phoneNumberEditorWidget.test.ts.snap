// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot phone number editor basic 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
}

.c2 {
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c7 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c7:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c7:hover,
.c7:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c7[disabled] {
  opacity: .65;
}

.c3 {
  position: relative;
}

.c5 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c4 {
  margin-bottom: 10px;
}

.c4 > label::after {
  content: "\\A0*";
  color: #dc4e41;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c6:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c6:disabled,
.c6[readonly] {
  background-color: #e9ecef;
}

.c6::-webkit-input-placeholder {
  color: #868e96;
}

.c6::-moz-placeholder {
  color: #868e96;
}

.c6:-ms-input-placeholder {
  color: #868e96;
}

.c6::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="c1 c2"
      >
        phoneNumberEditor.intro
      </div>
      <form
        class="c3"
        novalidate=""
      >
        <div
          class="c4"
          required=""
        >
          <label
            class="c5"
            for="phone_number"
          >
            phoneNumber
          </label>
          <input
            autocomplete="tel"
            class="c6"
            data-testid="phone_number"
            id="phone_number"
            labels="[object Object]"
            name="phone_number"
            placeholder="phoneNumber"
            required=""
            type="tel"
            value=""
          />
        </div>
        <button
          class="c7"
          data-testid="submit"
          type="submit"
        >
          send
        </button>
      </form>
    </div>
  </div>
</div>
`;
